### Installing and Importing Required Libraries

In [2]:
#Importing Required Libraries
import cvxpy as cp
import pandas as pd
import numpy as np

The regression equation is 
\begin{equation}
Profitability = 39.05 - (5.41 \times State\ Population\ Per\ Inn) + (5.86 \times Price\ Of\ Inn) - (3.09 \times Sqrt\ of\ the\ Median\ Income\ in\ the\ Area) + (1.75 \times College\ Students\ in\ the\ Area)
\end{equation}

### 1.
According to the regression equation, the variables that positively affect Profitability are 'Price of the Inn' and 'College Students in the Area'. On the other hand, 'State Population per Inn' and 'Square Root of the Median Income in the area' negatively affect Profitability.

This can intuitively make sense:

1. 'Price of the Inn' positively affects profitability. This is logical as a higher price for the inn's services would typically lead to higher revenue, and assuming costs are managed effectively, higher profitability.

2. The number of 'College Students in the Area' also positively affects profitability. This could be because college students may require accommodation for various reasons (e.g., visiting friends, attending events, parents visiting, etc.), and a higher number of college students could drive up demand for inn rooms, leading to higher utilization and profitability.

Conversely,

3. 'State Population per Inn' negatively impacts profitability. This could be interpreted as competition: if there are fewer people per inn, it might mean that the market is oversaturated with inns, and thus each inn has a smaller market share, leading to lower profitability.

4. The 'Square Root of the Median Income in the area' also negatively affects profitability. This might seem counterintuitive at first, as areas with higher income might be expected to have more disposable income for hotel stays. However, it could be that in higher-income areas, residents have the means to travel out of town rather than stay in local inns, or they might prefer more luxurious accommodations that a mid-sized hotel chain like LQM might not provide. Additionally, higher-income areas might have higher operating costs, which could negatively impact profitability.

### 2.

In [3]:
#Reading the data
lqm = pd.read_excel("LQM.xlsx")

#Setting column names to the values in the second row
lqm.columns = lqm.iloc[1]

#Dropping the first and second row
lqm = lqm.iloc[2:]

#Reseting the index
lqm.reset_index(drop = True, inplace = True)

#Creating a Profitability column based on the given formula
lqm['Profitability'] = 39.05 - (5.41 * lqm['State Population Per Inn (normalized)']) + (5.86 * lqm['Price (normalized)']) - (3.09 * lqm['Square Root of Median Income (normalized)']) + (1.75 * lqm['College Students in Area (normalized)'])

#Sorting by Profitability
lqm.sort_values(by = 'Profitability', ascending = False, inplace = True)

#Printing results
lqm

/usr/local/lib/python3.10/dist-packages/openpyxl/worksheet/_reader.py:312: UserWarning: Unknown extension is not supported and will be removed
  warn(msg)


1  Hotel                      Location     Price Price (normalized)  \
1      2            Fresno, California  10000000           1.699076   
5      6        Long Beach, California   8950000           1.402123   
0      1            Eureka, California   2925000          -0.301823   
2      3            Fresno, California   3750000          -0.068503   
3      4            Fresno, California   3500000          -0.139206   
13    14  South Lake Tahoe, California   3750000          -0.068503   
11    12  South Lake Tahoe, California   2500000          -0.422019   
12    13  South Lake Tahoe, California   1975000          -0.570495   
9     10  South Lake Tahoe, California   1650000          -0.662409   
14    15  South Lake Tahoe, California   1475000          -0.711902   
10    11  South Lake Tahoe, California   1125000          -0.810886   
15    16  South Lake Tahoe, California    750000          -0.916941   
4      5            Fresno, California    325000          -1.037136   
8      9       Los Angeles, California   4900000           0.256732   
6      7       Los Angeles, California   1950000          -0.577566   
7      8       Los Angeles, California   1750000          -0.634128   

1  Square Root of Median Income (normalized)  \
1                                  -0.408199   
5                                   0.657845   
0                                   -0.81278   
2                                  -0.408199   
3                                  -0.408199   
13                                 -0.791006   
11                                 -0.791006   
12                                 -0.791006   
9                                  -0.791006   
14                                 -0.791006   
10                                 -0.791006   
15                                 -0.791006   
4                                  -0.408199   
8                                   0.167686   
6                                   0.167686   
7                                   0.167686   

1  College Students in Area (normalized)  \
1                               0.311669   
5                               0.483711   
0                              -0.536413   
2                               0.311669   
3                               0.311669   
13                             -0.594905   
11                             -0.594905   
12                             -0.594905   
9                              -0.594905   
14                             -0.594905   
10                             -0.594905   
15                             -0.594905   
4                               0.311669   
8                               3.106215   
6                               3.106215   
7                               3.106215   

1  State Population Per Inn (normalized) Profitability  
1                              -0.474279     53.379192  
5                              -0.557278     49.095069  
0                              -0.995987     44.242369  
2                              -0.474279     43.021179  
3                              -0.474279     42.606858  
13                             -0.426851     42.360966  
11                             -0.426851     40.289363  
12                             -0.426851      39.41929  
9                              -0.426851     38.880673  
14                             -0.426851     38.590649  
10                             -0.426851       38.0106  
15                             -0.426851     37.389119  
4                              -0.474279     37.344988  
8                                3.10653     28.665848  
6                                3.10653     23.776866  
7                                3.10653     23.445409

We observe that Hotel 2 in Fresno, CA has the highest profitability and Hotel 8 in Los Angeles, CA has the lowest profitability.

### 3.

a. In the greedy approach, we start by purchasing the most profitable hotel and continue purchasing in descending order of profitability until we run out of budget. The hotels we purchase would be:

1. Fresno, California ($10,000,000) 
    - We would stop here because the total price ($10,000,000) matches our budget ($10,000,000).

b. Our total predicted profitability would be the sum of the profitability of the hotel we purchase, which is 53.379192 (from the hotel in Fresno, California).

c. If we are trying to maximize our total predicted profitability, this may not be the best approach. The greedy algorithm doesn't consider the cost-effectiveness of each hotel (profitability per dollar spent). A hotel might be less profitable overall, but more profitable per dollar spent, and thus a better investment. In this case, buying a single expensive hotel may not be the most cost-effective decision. 

If we were trying to maximize the average predicted profitability of the hotels we select, the approach would also need adjustments. It would be better to consider the profitability to cost ratio.

If we had a budget of \$20 million instead of \$10 million, the greedy approach would lead us to buy more hotels, starting from the most profitable until the budget is exhausted. However, the same considerations about cost-effectiveness and average profitability would apply. It would be better to consider the profitability per dollar spent on each hotel. We would en up buying Hotels 2 and 6 for a total price of \$18,950,000 and a total profitability of 102.469.

### 4.
a. The optimization problem can be formulated as follows:

We want to maximize the total profitability, subject to the constraint that the total price of the selected hotels does not exceed the budget.

Let's define the decision variables, objective function, and constraints.

Decision Variables:
Let's denote by $x_i$ a binary variable that is equal to 1 if hotel $i$ is selected, and 0 otherwise, where $i$ goes from 1 to $n$ (the number of hotels).

Objective Function:
We want to maximize total profitability, which can be written as:
\begin{equation}
\text{max}_{x}\quad \sum_{i=1}^{n}  \text{Profitability}_i \times x_i
\end{equation}

Constraints:
The total price of the selected hotels should not exceed the budget. This can be written as:
\begin{align}
\sum_{i=1}^{n}\text{Price}_i \times x_i &\leq \text{Budget}\\
x_i &\in \{0,1\}\quad \forall i
\end{align}



In [4]:
#Number of hotels
n = len(lqm)

#Decision variables - indicating whether a hotel is chosen
x = cp.Variable(n, boolean = True)

#Objective function
profit = lqm['Profitability'].values
objective = cp.Maximize(profit @ x)

#Constraints
budget = 10e6
costs = lqm['Price'].values
constraints = [costs @ x <= budget]

#Defining the problem
problem = cp.Problem(objective, constraints)

#Solving the problem
problem.solve()

#Observing the optimal solutions
chosen_hotels = x.value
optimal_profit = problem.value

for i in range(n):
    if chosen_hotels[i]:
        print(f"Hotel ID: {lqm.iloc[i]['Hotel']}, Location: {lqm.iloc[i]['Location']}, Price: {lqm.iloc[i]['Price']}")

print("\nOptimal Total Profitability: ",optimal_profit)

Hotel ID: 12, Location: South Lake Tahoe, California, Price: 2500000
Hotel ID: 13, Location: South Lake Tahoe, California, Price: 1975000
Hotel ID: 10, Location: South Lake Tahoe, California, Price: 1650000
Hotel ID: 15, Location: South Lake Tahoe, California, Price: 1475000
Hotel ID: 11, Location: South Lake Tahoe, California, Price: 1125000
Hotel ID: 16, Location: South Lake Tahoe, California, Price: 750000
Hotel ID: 5, Location: Fresno, California, Price: 325000

Optimal Total Profitability:  269.9246813771899


c. The optimal solution makes intuitive sense because it would select hotels to maximize profitability while staying within the budget. It might be different from the greedy solution because it considers the entire set of possible selections, rather than making one decision at a time. The optimal solution might select more lower-cost hotels with high profitability, rather than fewer high-cost hotels with slightly higher profitability. This illustrates the principle that considering the overall system can lead to better decisions than optimizing one decision at a time in isolation.

The optimal solution is to purchase 7 Hotels, ampunting to \$9,800,000 and a total profitability of 269.9.

### 5.
a. Constraints: To limit the number of hotels purchased in any city to at most two, we need to add a constraint for each city that the sum of the decision variables for the hotels in that city should be less than or equal to 2. We can define a binary matrix where each row corresponds to a city and each column corresponds to a hotel, with a value of 1 if the hotel is in the city and 0 otherwise. Multiplying this matrix by the decision variable vector and constraining each result to be less than or equal to 2 gives us the desired constraints.

Intuitively, the new optimal objective function value will be smaller or the same as before, because we are adding constraints which limit the feasible region of the problem. We can't expect the solution to be better, as we are restricting our options.

b.  The new optimization model:

\begin{equation}
\text{max}_{x}\quad \sum_{i=1}^{n}  \text{Profitability}_i \times x_i
\end{equation}

s.t.
\begin{align}
\sum_{i=1}^{n} \text{Price}_i \times x_{i} &\leq \text{Budget}\
\sum_{i=1}^{n} \text{CityMatrix}_{j,i} \times x_{i} \leq 2 \quad \forall j\ 
&\ x_{i} \in {0,1} \quad \forall i
\end{align}

In [5]:
#Getting a list of unique cities
cities = lqm['Location'].unique()

#Creating a binary matrix indicating which hotels are in which cities
city_matrix = np.zeros((len(cities), n))
for i, city in enumerate(cities):
    city_matrix[i] = (lqm['Location'] == city)

#Defining the Decision variables
x = cp.Variable(n, boolean=True)

#Defining the Objective function
profit = lqm['Profitability'].values
objective = cp.Maximize(profit @ x)

#Constraints
budget = 10e6
costs = lqm['Price'].values
constraints = [costs @ x <= budget]
for i in range(len(cities)):
    constraints.append(city_matrix[i] @ x <= 2)

#Defining the problem
problem = cp.Problem(objective, constraints)

#Solving the problem
problem.solve()

#Observing the optimal solutions
chosen_hotels = x.value
optimal_profit = problem.value

for i in range(n):
    if chosen_hotels[i]:
        print(f"Hotel ID: {lqm.iloc[i]['Hotel']}, Location: {lqm.iloc[i]['Location']}, Price: {lqm.iloc[i]['Price']}")

print("\nOptimal Total Profitability: ", problem.value)

Hotel ID: 1, Location: Eureka, California, Price: 2925000
Hotel ID: 10, Location: South Lake Tahoe, California, Price: 1650000
Hotel ID: 11, Location: South Lake Tahoe, California, Price: 1125000
Hotel ID: 5, Location: Fresno, California, Price: 325000
Hotel ID: 7, Location: Los Angeles, California, Price: 1950000
Hotel ID: 8, Location: Los Angeles, California, Price: 1750000

Optimal Total Profitability:  205.70090440379596


Compared to the previous solution, we observe that the optimal total profitability has decreased to 205.7 and the number of hotels recommended for purchase have gone down to 6 from 7. The total expenditure has decreased to \$9,725,000. 

This solution results in a smaller expenditure and recued profitability and hence would not be a recommended solution.


### 6.
I would begin by highlighting our main findings, which are the optimal selection of hotels under the given budget of $10 million. Our optimization model has been designed to maximize the total predicted profitability based on the regression model and we've considered both unrestricted and city-diversified approaches. The latter ensures that we're not overly invested in a single city by limiting the number of hotels acquired in any city to a maximum of two.

However, I would recommend that LQM consider refining their regression model to potentially include other influential variables such as proximity to tourist attractions, availability of amenities, and competition in the area. This would allow for a more comprehensive analysis of hotel profitability. I would also suggest incorporating interaction terms if there are dependencies between variables, and examining potential nonlinear relationships to improve the model's accuracy.

For the optimization model, while the greedy approach and constrained approach provide good solutions, we might explore a multi-objective optimization that balances profitability with other factors such as risk diversification, growth opportunities, and strategic fit. Furthermore, it might be beneficial to consider a stochastic optimization approach if there are uncertainties in the predicted profitability or budget.

Overall, our analysis should serve as a valuable starting point, but a more refined model could yield even more effective strategies for LQM's expansion plans.